### PR TITLE
Matt Simon

### PR DESCRIPTION
**Pull request by msimon42**

### CODE REVIEW

File: lib/house.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+require_relative 'room'
+
+class House
+  attr_reader :price, :address, :rooms
+
+  def initialize(price, address, rooms=Array.new)
```
Comment (aziobrow) [inline lib/house.rb]: This isn't bad in terms of setting a default value for rooms, but since we never see any occasion in which a house is created with rooms, setting this default value is better done inside of `initialize`.

File: lib/house.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @address = address
+    @rooms = rooms
+  end
+
+  def add_room(room)
+    @rooms << room
+  end
+
+  def rooms_from_category(category)
+    @rooms.find_all {|room| room.category == category.to_sym}
```
Comment (aziobrow) [inline lib/house.rb]: nice choice of find_all, good defensive coding with coercing the category to symbol.

File: lib/house.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def area
+    @rooms.collect { |room| room.area }.reduce(0, :+)
```
Comment (aziobrow) [inline lib/house.rb]: Nice effort to use reduce!

Since reduce takes a block, you actually don't need to collect first, you can do this:

`@rooms.reduce(0) {|sum, room| sum + room.area}`

This is preferred so that we're not iterating twice unecessarily.

File: lib/house.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def price_per_sf
+    (self.price.to_f / self.area.to_f).round(2)
```
Comment (aziobrow) [inline lib/house.rb]: only one of these needs to be changed to a float, although doing both doesn't hurt.

File: lib/house.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def sort_by_area
+    @rooms.sort_by {|room| room.area}
```
Comment (aziobrow) [inline lib/house.rb]: nice!

File: lib/house.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def rooms_by_category
+    categories = @rooms.uniq {|room| room.category}.map {|room| room.category}
```
Comment (aziobrow) [inline lib/house.rb]: The uniq needs to go on the end-- @rooms is an array of room objects, which should already be unique (unless you added the same room more than once, for some reason, but that's unlikely).  You 'd want to map to collect all of the categories first, and then you'd have something like this: `[:bedroom, :bedroom, :dining_room, :basement]`, which you'd then call uniq on (to get rid of the duplication of the bedroom category)

File: lib/house.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    result = Hash[categories.collect {|cat| [cat, Array.new]}]
+    categories.each do |cat|
+      result[cat] = self.rooms_from_category(cat)
+    end
+    result
+  end
```
Comment (aziobrow) [inline lib/house.rb]: Really solid work figuring this out--this is some pretty complex stuff.  I'm sure you saw this during your walkthrough, but `group_by` is an interesting tool (http://gregpark.io/blog/ruby-group-by/) which actually behaves differently, depending on the data type it's called on.  In this case, it gives you exactly what you need.  But great job figuring out an alternative path, with more complex logic.

File: lib/room.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  def initialize(category, length, width)
+    @category = category
+    @length = length
+    @width = width
+  end
+
+  def area
+    @length * @width
+  end
+end
```
Comment (aziobrow) [inline lib/room.rb]: 👍

File: test/house_test.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    assert_equal [@room1, @room2, @room3, @room4], @house.rooms
+  end
+
+  def test_find_rooms_from_cat
+    @house.add_room(@room1)
+    @house.add_room(@room2)
+    @house.add_room(@room3)
+    @house.add_room(@room4)
+
+    assert_equal [@room1, @room2], @house.rooms_from_category('bedroom')
```
Comment (aziobrow) [inline test/house_test.rb]: i like that you're edge case testing if it works with a string-- make sure it still works with a symbol too, and it's probably best to check more than one category

File: lib/house.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+  def rooms_from_category(category)
+    @rooms.find_all {|room| room.category == category.to_sym}
+  end
+
+  def area
+    @rooms.collect { |room| room.area }.reduce(0, :+)
+  end
+
+  def price_per_sf
```
Comment (aziobrow) [inline lib/house.rb]: make sure your naming convention exactly matches the interaction pattern-- remember, that's essentially the business criteria that you're being asked to fulfill. 

```
pry(main)> house.price_per_square_foot
#=> 210.53
```

File: test/house_test.rb [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    assert_equal 366.64, @house.price_per_sf
+  end
+
+  def test_sort_by_area
+    @house.add_room(@room1)
+    @house.add_room(@room2)
+    @house.add_room(@room3)
+    @house.add_room(@room4)
+
+    assert_equal [@room1, @room4, @room2, @room3], @house.sort_by_area
```
Comment (aziobrow) [inline test/house_test.rb]: Again, follow interaction pattern really closely for exact order that we're expecting.

File: test/house_test.rb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  end
+
+  def test_rooms_by_cat
+    @house.add_room(@room1)
+    @house.add_room(@room2)
+    @house.add_room(@room3)
+    @house.add_room(@room4)
+
+    hash = {:bedroom=>[@room1, @room2], :living_room => [@room3], :kitchen => [@room4]}
+    assert_equal hash, @house.rooms_by_category
```
Comment (aziobrow) [inline test/house_test.rb]: Nice use of a variable for expected value to shorten things up and to make it easier to read .We typically don't name variables their data types, so instead of calling this `hash`, maybe call it `grouped_rooms`, or even something more generic like `expected_value`.
